File: Dart/merge.dart
void main() {
  var arr = [10, 20, 3, 6, 1, 0];
  
  print(arr);
  Stopwatch stopwatch = new Stopwatch()..start();
  print(merge(arr));
  print('Sort executed in ${stopwatch.elapsed}');
}

List merge(arr){
  if(arr.length > 1){
    var L = arr.length;
    var a1 = arr.sublist(0, L ~/ 2);
    var a2 = arr.sublist(L ~/ 2, L);
    a1 = merge(a1);
    a2 = merge(a2);
    var i = 0;
    var j = 0;
    var k = 0;
    while(i < a1.length && j < a2.length){
      if(a1[i] < a2[j]){
        arr[k] = a1[i];
        i++;
      }
      else{
        arr[k] = a2[j];
        j++;
      }
      k++;
    }
    while(i < a1.length){
      arr[k] = a1[i];
      i++;
      k++;
    }
    while(j < a2.length){
      arr[k] = a2[j];
      j++;
      k++;
    }
  }
  return arr;
}

File: Rust/merge.rs
use std::time::{Instant};

fn main() {
	let mut vec = vec![10, 20, 3, 6, 1, 0];
	let L = vec.len();

	for x in &vec {
		print!("{} ", x);
	}

	let start = Instant::now();
	vec = merge(vec.clone());
	let duration = start.elapsed();

	println!("");
	for x in &vec {
		print!("{} ", x);
	}

	println!("\nTime elapsed in expensive_function() is: {:?}", duration);
}

fn merge(mut vec: Vec<i32>) -> Vec<i32>{
	if(vec.len() > 1){
		let mut v1 = vec![];
		let mut v2 = vec![];
		for i in 0..vec.len(){
			if(i < vec.len() / 2){
				v1.push(vec[i])
			}
			else{
				v2.push(vec[i])
			}
		}
		v1 = merge(v1);
		v2 = merge(v2);
		let mut i = 0;
		let mut j = 0;
		let mut k = 0;
		while(i < v1.len() && j < v2.len()){
			if(v1[i] < v2[j]){
				vec[k] = v1[i];
				i += 1;
			}
			else{
				vec[k] = v2[j];
				j += 1;
			}
			k += 1;
		}
		while(i < v1.len()){
			vec[k] = v1[i];
			i += 1;
			k += 1;
		}
		while(j < v2.len()){
			vec[k] = v2[j];
			j += 1;
			k += 1;
		}
	}
	return vec;
}